src/utils/xhr-utils.js

```diff
@@ -0,0 +1,626 @@
+  const whenFirstVisible = viewer.whenFirstVisible();
+  if (isProxyOrigin(input) || !viewer.hasCapability('xhrInterceptor')) {
+    return whenFirstVisible;
+  }
+  const htmlElement = ampdocSingle.getRootNode().documentElement;
+  const docOptedIn = htmlElement.hasAttribute('allow-xhr-interception');
+  if (!docOptedIn) {
+    return whenFirstVisible;
+  }
+  return whenFirstVisible.then(() => {
+    return viewer.isTrustedViewer();
+  }).then(viewerTrusted => {
+    const isDevMode = getMode(win).development;
+    if (!viewerTrusted && !isDevMode) {
+      return;
+    }
+    const messagePayload = dict({
+      'originalRequest': toStructuredCloneable(input, init),
+    });
+    return viewer.sendMessageAwaitResponse('xhr', messagePayload)
+        .then(response =>
+          fromStructuredCloneable(response, init.responseType));
+  });
+}
+
+/**
+ * Setsup URL based on ampCors
+ * @param {!Window} win
+ * @param {string} input
+ * @param {!FetchInitDef} init The options of the XHR which may get
+ * intercepted.
+ */
+export function setupInput(win, input, init) {
+  dev().assert(typeof input == 'string', 'Only URL supported: %s', input);
+  if (init.ampCors !== false) {
+    input = getCorsUrl(win, input);
+  }
+  return input;
+}
+
+/**
+ * Sets up and normalizes the FetchInitDef
+ *
+ * @param {?FetchInitDef=} opt_init Fetch options object.
+ * @param {string=} opt_accept The HTTP Accept header value.
+ * @return {!FetchInitDef}
+ */
+export function setupInit(opt_init, opt_accept) {
+  const init = opt_init || {};
+
+  // In particular, Firefox does not tolerate `null` values for
+  // `credentials`.
+  const creds = init.credentials;
+  dev().assert(
+      creds === undefined || creds == 'include' || creds == 'omit',
+      'Only credentials=include|omit support: %s', creds);
+
+  init.method = normalizeMethod_(init.method);
+  init.headers = init.headers || dict({});
+  if (opt_accept) {
+    init.headers['Accept'] = opt_accept;
+  }
+
+  return init;
+}
+
+/**
+ *
+ * Sets up AMPSpecific CORS headers.
+ * @param {!Window} win
+ * @param {string} input
+ * @param {?FetchInitDef=} init
+ * @return {!FetchInitDef}
+ */
+export function setupAMPCors(win, input, init) {
+  // Do not append __amp_source_origin if explicitly disabled.
+  if (init.ampCors === false) {
+    init.requireAmpResponseSourceOrigin = false;
+  }
+  if (init.requireAmpResponseSourceOrigin === true) {
+    dev().error('XHR',
+        'requireAmpResponseSourceOrigin is deprecated, use ampCors instead');
+  }
+  if (init.requireAmpResponseSourceOrigin === undefined) {
+    init.requireAmpResponseSourceOrigin = true;
+  }
+  // For some same origin requests, add AMP-Same-Origin: true header to allow
+  // publishers to validate that this request came from their own origin.
+  const currentOrigin = getWinOrigin(win);
+  const targetOrigin = parseUrlDeprecated(input).origin;
+  if (currentOrigin == targetOrigin) {
+    init['headers'] = init['headers'] || {};
```

Comment:
Nit: This should be [guaranteed](https://github.com/ampproject/amphtml/pull/17206/files#diff-6870dab961227bda821a287600fb7158R281).